src/bustools_project.cpp

```diff
@@ -0,0 +1,169 @@
+#include "BUSData.h"
+
+#include "bustools_project.h"
+
+void bustools_project(Bustools_opt &opt) {
+  BUSHeader h;
+  std::ofstream of;
+
+  /* Read and parse equivalence class files. */
+  std::unordered_map<std::string, int32_t> txnames;
+  parseTranscripts(opt.count_txp, txnames);
+
+  std::vector<int32_t> genemap(txnames.size(), -1);
+  std::unordered_map<std::string, int32_t> genenames;
+  parseGenes(opt.count_genes, txnames, genemap, genenames);
+  std::vector<std::string> genenamesinv(genenames.size(), "");
+  for (const auto &gene : genenames) {
+    genenamesinv[gene.second] = gene.first;
+  }
+
+  std::unordered_map<std::vector<int32_t>, int32_t, SortedVectorHasher> ecmapinv;
+  std::vector<std::vector<int32_t>> ecmap;
+  parseECs(opt.count_ecs, h);
+  ecmap = std::move(h.ecs);
+  ecmapinv.reserve(ecmap.size());
+  for (int32_t ec = 0; ec < ecmap.size(); ec++) {
+    ecmapinv.insert({ecmap[ec], ec});
+  }
+
+  std::vector<std::vector<int32_t>> ec2genes;        
+  create_ec2genes(ecmap, genemap, ec2genes);
+
+  std::vector<std::vector<int32_t>> geneEc2genes = ec2genes;
+  std::unordered_map<std::vector<int32_t>, int32_t, SortedVectorHasher> geneEc2genesinv;
+  std::sort(geneEc2genes.begin(), geneEc2genes.end());
+  auto firstNonempty = geneEc2genes.begin();
+  while (firstNonempty->size() == 0 && firstNonempty != geneEc2genes.end()) {
+    ++firstNonempty;
+  }
+  geneEc2genes.erase(geneEc2genes.begin(), firstNonempty);
+  geneEc2genes.erase(std::unique(geneEc2genes.begin(), geneEc2genes.end()), geneEc2genes.end());
+  for (int32_t ec = 0; ec < geneEc2genes.size(); ++ec) {
+    geneEc2genesinv.insert({geneEc2genes[ec], ec});
+  }
+  geneEc2genesinv.insert({{}, -1});
+
+  std::vector<int32_t> txEc2geneEc;
+  for (const auto &txEc : ec2genes) {
+    txEc2geneEc.push_back(geneEc2genesinv.at(txEc));
+  }
+  
+
+  /* Write gene EC matrix. */
+  of.open(opt.output + ".ec");
+  for (int i = 0; i < geneEc2genes.size(); ++i) {
+    std::string geneEc = "";
+    for (const auto &gene : geneEc2genes[i]) {
+      geneEc += ",";
+      geneEc += std::to_string(gene);
+    }
+    of << i << '\t' << geneEc.substr(1) << '\n';
+  }
+  of.close();
+
+  
+  /* Write gene table. */
+  of.open(opt.output + ".genes.txt");
+  for (const auto & gene : genenamesinv) {
+    of << gene << '\n';
+  }
+  of.close();
+
+
+  /* Open input/output BUS files and write header. */
+
+  std::streambuf *buf = nullptr;
+  if (!opt.stream_out) {
+    of.open(opt.output + ".bus"); 
+    buf = of.rdbuf();
+  } else {
+    buf = std::cout.rdbuf();
+  }
+  std::ostream o(buf);
+
+  std::streambuf *inbuf;
+  std::ifstream inf;
+  if (!opt.stream_in) {
+    inf.open(opt.files[0].c_str(), std::ios::binary);
+    inbuf = inf.rdbuf();
+  } else {
+    inbuf = std::cin.rdbuf();
+  }
+  std::istream in(inbuf);
+  parseHeader(in, h);
+
+  h.transcripts.clear();
+  for (const auto & gene : genenamesinv) {
+    h.transcripts.emplace_back(gene);
+  }
+
+  h.ecs = geneEc2genes;
```

Comment:
Use move semantics to avoid unnecessary copy, also clears out `geneEc2genes` afterwards so make sure it's not being used below

```suggestion
  h.ecs = std::move(geneEc2genes);
```